keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,519 @@
+
+![API Server Network Proxy Simple Cluster](NetworkProxySimpleCluster.png)
+The diagram shows API Server’s outgoing traffic flow.
+The user (in blue box), master network (in purple cloud) and
+a cluster network (in green cloud) are represented.
+
+The user (blue) initiates communication to the KAS.
+The KAS then initiates connections to other components.
+It could be node/pod/service in cluster networks (red dotted arrow to green cloud),
+or etcd for storage in the same master network (blue arrow) or mutate the request
+based on an admission web-hook (red dotted arrow to purple cloud).
+The KAS handles these cases based on NetworkContext based traffic routing.
+The connectivity proxy should be able to do routing solely based on IP.
+The proxy should not require the NetworkContext. This means the service CIDR,
+node CIDR and pod CIDR of each cluster network cannot overlap.
+
+### Network Context
+
+The minimal NetworkContext looks like the following struct in golang:
+
+```go
+type NetworkContext struct {
+        // ConnectivityServiceName is the unique name of the
+        // ConnectivityServiceConfiguration which determines
+        // the network we route the traffic to.
+        ConnectivityServiceName string
+}
+```
+
+ConnectivityServiceName specifies the network to route traffic to.
+The KAS starts with a list of registered connectivity service names. These
+correspond to networks we route traffic to. So the KAS knows where to
+proxy the traffic to, otherwise it return an “Unknown network” error.
+
+The KAS starts with a proxy configuration like the below example.
+The example specifies 4 networks. "direct" specifies the KAS talking directly
+on the local network (no proxy). "master" specifies the KAS talks to a proxy
+listening at 1.2.3.4:5678. "cluster" specifies the KAS talk to a proxy
+listening at 1.2.3.5:5679. While these are represented as resources
+they are not intended to be loaded dynamically.
+The KAS loads them as configuration at start time.
+
+```yaml
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: direct
+  connection:
+    type: direct
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Master
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5678
+    caBundle: file1.pem
+    clientKeyFile: proxy-client1.key
+    clientCertFile: proxy-client1.crt
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Cluster
+  connection:
+    type: grpc
+    url: grpc://1.2.3.5:5679
+    caBundle: file2.pem
+    clientKeyFile: proxy-client2.key
+    clientCertFile: proxy-client2.crt
+```
+
+NetworkContext could be extended to contain more contextual information.
+This would allow smarter routing based on the k8s object KAS is processing
+or which user/tenant tries to initiate the request, etc.
+
```

Comment:
How are we going to do load balancing api-server -> **connectivity proxy** ?
Option 1: Urls becomes a list, users can run the proxy on per master node bases or 2 or just one. That means the api server needs to maintain `rr` state of last connection
Option 2: Forcing the user to run an internal load balancer.

I vote `option 1`

Reply:
never mind. below you are asserting same node as api-server.